probe-rs/src/probe/blackmagic/mod.rs

```diff
@@ -0,0 +1,1554 @@
+    }
+}
+
+/// Determine if a given serial port is a Black Magic Probe GDB interface.
+/// The BMP has at least two serial ports, and we want to make sure we get
+/// the correct one.
+fn black_magic_debug_port_info(
+    port_type: SerialPortType,
+    port_name: &str,
+) -> Option<DebugProbeInfo> {
+    // Only accept /dev/cu.* values on macos, to avoid having two
+    // copies of the port (both /dev/tty.* and /dev/cu.*)
+    if cfg!(target_os = "macos") && !port_name.contains("/cu.") {
+        return None;
+    }
+
+    let (vendor_id, product_id, serial_number, hid_interface, identifier) = match port_type {
+        SerialPortType::UsbPort(info) => (
+            info.vid,
+            info.pid,
+            info.serial_number.map(|s| s.to_string()),
+            info.interface,
+            info.product
+                .unwrap_or_else(|| "Black Magic Probe".to_string()),
+        ),
+        #[cfg(target_os = "linux")]
+        SerialPortType::Unknown => {
+            use std::{
+                fs::File,
+                path::{Path, PathBuf},
+            };
+
+            let port_name = port_name.split('/').last()?;
+            let device_path = PathBuf::from(format!("/sys/class/tty/{}/device", port_name))
+                .canonicalize()
+                .ok()?;
+            let subsystem = device_path.join("subsystem").canonicalize().ok()?;
+            let subsystem = subsystem.file_name()?.to_string_lossy();
+
+            let usb_interface_path = if subsystem == "usb-serial" {
+                device_path.parent()?
+            } else if subsystem == "usb" {
+                &device_path
+            } else {
+                return None;
+            };
+            let usb_device_path = usb_interface_path.parent()?;
+
+            fn read_file_to_string(dir: &Path, file: &str) -> Option<String> {
+                let dir = dir.join(file);
+                let mut f = File::open(dir).ok()?;
+                let mut s = String::new();
+                f.read_to_string(&mut s).ok()?;
+                Some(s.trim().to_owned())
+            }
+
+            fn read_file_to_u16(dir: &Path, file: &str) -> Option<u16> {
+                u16::from_str_radix(&read_file_to_string(dir, file)?, 16).ok()
+            }
+
+            fn read_file_to_u8(dir: &Path, file: &str) -> Option<u8> {
+                u8::from_str_radix(&read_file_to_string(dir, file)?, 16).ok()
+            }
+
+            let vid = read_file_to_u16(&usb_device_path, &"idVendor")?;
+            let pid = read_file_to_u16(&usb_device_path, &"idProduct")?;
+            let interface_number = read_file_to_u8(&usb_interface_path, &"bInterfaceNumber");
+            let serial_number = read_file_to_string(&usb_device_path, &"serial");
+            let product_name = read_file_to_string(&usb_device_path, &"product")
+                .unwrap_or_else(|| "Black Magic Probe".to_string());
+            (vid, pid, serial_number, interface_number, product_name)
+        }
```

Comment:
Is that the part that would be PR'd into serialport? If so, reviewers might appreciate a note right in the code; if that PR already exists, maybe even pointing to that PR so that it is easy to simplify the code again once serialport has improved.

(I don't think that having the workaround in here is detrimental to inclusion in probe-rs, especially if it is a clearly indicated workaround with a path forward, and the workaround is necessary due to probe-rs's policy not to use libudev).

Reply:
That's a good suggestion! I've added a comment to that effect and linked to the PR.

The new version isn't out yet, and there's no word on when it will be out. But once they release it, the workaround will continue to function and can be removed at a later date.